omroneipApp/src/drvOmroneip.cpp

```diff
@@ -0,0 +1,1882 @@
+          }
+          else {
+            parentNextDtype ="none";
+          }
+          int structSize = findOffsetsRecursive(rawMap, expandedMap, dtype, structMap, parentNextDtype);
+          if (structSize <= 0)
+          {
+            structInvalid = true;
+            std::cout<<"Error calculating the size of structure: " << dtype << ". Definition for " << structName << " is invalid" << std::endl; 
+            return -1;
+          }
+          thisOffset = structSize+lastOffset;
+        }
+        catch (...)
+        {
+          structInvalid = true;
+          std::cout<<"Failed to find the standard datatype: " << dtype << ". Definition for " << structName << " and its dependents failed" << std::endl;
+          return -1;
+        }
+      }
+    }
+    if (offSetsCounted < row.size())
+    {
+      newRow.push_back(thisOffset);
+    }
+    else
+    {
+      // if this is the last offset, then do not add it to the row as it points to the end of the struct
+      // instead we return thisOffset
+      if (!structInvalid)
+      {
+        structMap[structName] = newRow;
+        return thisOffset;
+      }
+      else
+        return -1;
+    }
+  }
+  // Dont think its possible to get here
+  return -1;
+}
+
+std::string drvOmronEIP::findNextDtype(std::unordered_map<std::string, std::vector<std::string>> const& rawMap, std::string structName)
+{
+  try
+  {
+    std::vector<std::string> row = rawMap.at(structName);
+    for (std::string dtype : row)
+    {
+      if (dtype =="UINT" || dtype =="INT" || dtype =="WORD" || dtype =="BOOL" || dtype =="UDINT" || dtype =="DINT" || dtype =="REAL" || dtype =="DWORD" || dtype =="ULINT" || dtype =="LINT" || dtype == "LREAL" || dtype =="LWORD" || dtype.substr(0,6) == "STRING") {
+        return dtype;
+      }
+      else {
+        // We need to find the next dtype within an embedded structure
+        std::string nextDtype = findNextDtype(rawMap, dtype);
+        return nextDtype;
+      }
+    }
+  }
+  catch (...)
+  {
+    if (structName =="UINT" || structName =="INT" || structName =="WORD" || structName =="BOOL" || structName =="UDINT" || structName =="DINT" || structName =="REAL" || structName =="DWORD" || structName =="ULINT" || structName =="LINT" || structName == "LREAL" || structName =="LWORD" || structName.substr(0,6) == "STRING" ){
+      return structName;
+    }
+    else {
+      return "structName is neither a structure or a standard dtype";
+    }
+  }
+  return "no dtypes left";
+}
+
+void drvOmronEIP::readPoller()
+{
+  static const char *functionName = "readPoller";
+  std::string threadName = epicsThreadGetNameSelf();
+  std::string tag;
+  int offset;
+  int status;
+  int still_pending = 1;
+  int timeTaken = 0;
+  //The poller may not be fully initialised until the startPollers_ flag is set to 1, do not attempt to get the pPoller yet
+  while (!this->startPollers_ && !omronExiting_) {epicsThreadSleep(0.1);}
```

Comment:
I eventually caught up with Alan here... I agree with this one. I would definitely use epicsEvents over while(sleep) loops.

Reply:
Each configured poller runs in its own thread which runs the readPoller() function. So if I changed this to an event and triggered it when the driver is ready to start the pollers, only the first polling thread which checks for the event will break out of the loop. I could then use this poller to refire the event and so on until all pollers have seen it, but this would be a bit messy. Hence why I have used this flag instead which is set to true in one place only, and is checked by all of the different polling threads.